### PR TITLE
Add CRUD methods to public REST API

### PR DESCRIPTION
**Pull request by marcelklehr**
Hi,

after my frustrations with the current API (while working on https://github.com/marcelklehr/floccus) I decided to extend the public API with full CRUD support. The methods are mostly copied from the Bookmark controller, as the framework is new to me, so please excuse any bugs :)

### CODE REVIEW

**Comment by marcelklehr** [comment]
This is a new take on #218, as that is almost a year old now. No tests, yet. Please provide feedback.

**Comment by marcelklehr** [comment]
(I made the same mistake as the last PR. Now, this PR lets nc handle auth, but still disables CSRF, since it's an API, not a website.)

**Comment by blizzz** [comment]
@marcelklehr cool, thanks for your contribution! I will have a look in detail, nevertheless this is really great! Meanwhile, the tests are failing, could you have a look?

**Comment by marcelklehr** [comment]
Scrutinizer code coverage timed out, some test configurations on travis error with "App "Bookmarks" cannot be installed because it is not compatible with this version of the server." Not sure what that's about.

**Comment by marcelklehr** [comment]
There's still one failing test, I don't get any output or error message though :/

(PS: Sorry for the noise; squashing recommendedTM ;) )

**Comment by blizzz** [comment]
> Scrutinizer code coverage timed out, 

I saw all red yesterday, perhaps this was it then.

> some test configurations on travis error with "App "Bookmarks" cannot be installed because it is not compatible with this version of the server." Not sure what that's about.

Nextcloud master went to 12, which according to our info.xml is not supported, yet. It will also fail, because we are using IDb, which was already deprecated and kicked out. I am working on  replacing that.

Let me deal with making  Nextcloud master work again,  then we can rebase and look further, OK?

**Comment by marcelklehr** [comment]
Ah, ok. Good thing the public controller only uses libbookmarks :)

**Comment by marcelklehr** [comment]
I'm unable to fix the tests atm, it seems. @blizzz you might be quicker in finding the bug. Delete seems to work, so I'm at a loss.. :)

**Comment by blizzz** [comment]
Thanks @marcelklehr so far. I am going to have a look on it, but not sure whether I manage it today.

**Comment by marcelklehr** [comment]
Does testuser actually exist? Looks like it doesn't: https://travis-ci.org/nextcloud/bookmarks/jobs/185447600#L388

**Comment by marcelklehr** [comment]
Phew. Got sucked in way deeper than I wanted to, but it works now. Also I apologize for the spam commits :)

**Comment by Yamakasi** [comment]
Would be great to have this in again it's a missing feature! I hope to see it merged asap :) Thanks!

File: controller/rest/publiccontroller.php [hunk 1]
```diff
@@ -12,15 +13,120 @@
...
+	 * @param string $title
+	 * @param bool $is_public
+	 * @param string $description
+	 * @return JSONResponse
+	 *
+	 * @CORS
+	 * @NoAdminRequired
+	 * @NoCSRFRequired
+	 */
+	public function newBookmark($url = "", $item = array(), $title = "", $is_public = false, $description = "") {
```
Comment (blizzz) [inline controller/rest/publiccontroller.php]: please update the PHP Doc accordingly ↑

Comment (blizzz) [inline controller/rest/publiccontroller.php]: Hm, this seems to be the same code as in BookmarkController::newBookmark(), isn't it? We should not duplicate the code, but have it in one place. lib/bookmarks.php could have this perhaps, the instance is already available to both PublicController and BookmarkController.

File: controller/rest/publiccontroller.php [hunk 2]
```diff
@@ -12,15 +13,120 @@
...
+	 * @param null $record_id
+	 * @param string $description
+	 * @return JSONResponse
+	 *
+	 * @CORS
+	 * @NoAdminRequired
+	 * @NoCSRFRequired
+	 */
+	public function editBookmark($record_id = null, $url = "", $item = array(), $title = "", $is_public = false, $description = "") {
+		// Check if it is a valid URL
```
Comment (blizzz) [inline controller/rest/publiccontroller.php]: same as the comment for  ``newBookmark``

File: controller/rest/publiccontroller.php [hunk 3]
```diff
@@ -12,15 +13,120 @@
...
+	
+	/**
+	 * @param int $id
+	 * @return \OCP\AppFramework\Http\JSONResponse
+	 *
+	 * @CORS
+	 * @NoAdminRequired
+	 * @NoCSRFRequired
+	 */
+	public function deleteBookmark($id = -1) {
```
Comment (blizzz) [inline controller/rest/publiccontroller.php]: same as the comment for  ``newBookmark``

File: controller/rest/publiccontroller.php [hunk 4]
```diff
@@ -95,4 +201,16 @@ public function newJsonErrorMessage($message) {
 		return new JSONResponse($output);
 	}
 
+	/**
+	 * Checks whether parse_url was able to return proper URL data
+	 *
+	 * @param bool|array $urlData result of parse_url
+	 * @return bool
+	 */
+	protected function isProperURL($urlData) {
```
Comment (blizzz) [inline controller/rest/publiccontroller.php]: and this is also duplicated

**Comment by marcelklehr** [comment]
@blizzz Thanks for the feedback.
I agree that drier code is good. In this case I wasn't sure how to move that stuff somewhere else. One option would be to simply reference methods of the the bookmarks controller. Moving that stuff into libbookmarks would require differentiation of the different failure cases, (possibly) creating new Exceptions and handling them in the controllers, as libBookmarks probably shouldn't create JSONResponse instances directly. I'd say that's your call :)

**Comment by blizzz** [comment]
@marcelklehr OK, then let's merge it and I will do a subsequent PR with the consolidation. By doing so I will also remove the ridiculous user and password parameters from the GET request (which was there before).

Another thing we could think about is to use the website URL instead of the numeric iD for edit and deletion. Just came to my mind.

**Comment by blizzz** [comment]
Oh and this qualifies to increase versioning to 0.10 I think :) @marcelklehr thanks again for your contribution

**Comment by marcelklehr** [comment]
You're welcome -- it wasn't entirely selfless :) Thanks for merging!

re user and password parameters: I left those in, since they have a usecase: If you don't provide a password you'll get the public bookmarks of user x. Might be worth splitting both use cases into two endpoints, though. I'm not sure.

**Comment by blizzz** [comment]
Yes, you could specify a "public" and a "user" parameter, but splitting this up is better I think. Anyhow, there is no built-in way to create public bookmarks. Well, until this PR :)

**Comment by marcelklehr** [comment]
Come to think about it, bookmarkcontroller could probably be fully replaced by the public controller, can it not?

**Pull request merged**